protocol/opcm-v2.md

```diff
@@ -0,0 +1,234 @@
+- `deploy(Config memory config)`: Deploys a new OP Stack chain
+- `upgrade(Config memory config)`: Upgrades an existing chain
+
+Both functions share the same initialization path for contracts, eliminating the special upgrade functions that previously required maintenance between releases.
+
+### Key Architectural Changes
+
+**1. Dynamic Dispute Game Interface**
+
+Replace hardcoded assumptions about game types with dynamic arrays:
+```solidity
+struct GameConfig {
+    GameType gameType;
+    address implementation;
+    bytes32 prestate;
+}
+
+struct Config {
+    // ... other config
+    GameConfig[] games;  // Support arbitrary number of game types
+}
+```
+
+This eliminates the need for `addGameType()` and `updatePrestate()` as separate functions—these operations become implicit in `upgrade()` when the games array differs from the current state.
+
+**2. Unified Initialization Pattern**
+
+Both `deploy()` and `upgrade()` use the same contract initialization path:
+- Clear the `initialized` slot (for upgrades)
+- Call the standard `initialize()` function
+- No special upgrade functions that need removal in subsequent releases
+
+This ensures:
+- Single code path for both operations
+- Always release-ready (no post-upgrade cleanup)
+- Easier testing and maintenance
+
+**3. Auto-Generated Integration Code**
+
+Generate the Go glue code between OPCM and op-deployer from Solidity interfaces:
+- Solidity function signatures/ABIs are the source of truth
+- Go code for constructing function calls is automatically generated
+- Developers only need to add new CLI arguments to op-deployer when introducing net-new inputs
+- Reduces knowledge burden—developers don't need to understand integration layer
+
+### Implementation Flow
+
+Both `deploy()` and `upgrade()` follow a unified three-phase flow:
+
+**Phase 1: Chain World Assembly (Load-or-Deploy Pattern)**
+
+Both functions begin by gathering the complete set of contracts for the chain—the "chain world"—which includes SystemConfig, OptimismPortal, AnchorStateRegistry, and all other chain contracts.
+
+The load-or-deploy pattern attempts to load each contract address from an expected source (e.g., a known deployment address for upgrades, or a predictable location). If the contract cannot be found at the expected location, it is deployed instead. This unified pattern means the same chain world loading function serves both deploy and upgrade operations.
+
+**Phase 2: Input Gathering**
+
+After assembling the chain world, both functions gather the configuration inputs needed to properly configure the proxy contracts:
+
+- **Deployments**: All configuration input comes directly from the deploy input parameter
+- **Upgrades**: Only a subset of configuration is provided in the upgrade input; the remaining configuration is gathered automatically from the existing deployed contracts
+
+This asymmetry allows upgrades to specify only what changes, while deployments must provide complete configuration.
```

Comment:
How is this expressed in code? How do you know what you need to provide when calling upgrade?  The ABIs above pass the same Config object to both deploy and upgrade so it seems unclear how to know if an upgrade requires a new setting be specified or not.

Reply:
I think what's confusing me is lines 83 and 84 have `deploy(Config)` and `upgrade(Config)` which makes it look like they take the same inputs. In which case, how do you know which parts of that `Config` are required to be specified for the upgrade being done?

https://github.com/ethereum-optimism/design-docs/pull/353/files/1acc722cdf458c0744e166c0c37443a4a8b3f3b4#diff-62c8771657b68fe38ae394ef3d255c7928c0b7853b0077ca70ff62007de8beeeR83

But maybe I'm looking at the wrong place for the ABI you updated.